src/cargo/core/workspace.rs

```diff
@@ -1841,3 +1845,69 @@ impl InheritableFields {
             })
     }
 }
+
+fn parse_manifest(manifest_path: &Path, config: &Config) -> CargoResult<EitherManifest> {
+    let key = manifest_path.parent().unwrap();
+    let source_id = SourceId::for_path(key)?;
+    let (manifest, _nested_paths) = read_manifest(manifest_path, source_id, config)?;
+    Ok(manifest)
+}
+
+pub fn find_workspace_root(manifest_path: &Path, config: &Config) -> CargoResult<Option<PathBuf>> {
+    fn read_root_pointer(member_manifest: &Path, root_link: &str) -> PathBuf {
+        let path = member_manifest
+            .parent()
+            .unwrap()
+            .join(root_link)
+            .join("Cargo.toml");
+        paths::normalize_path(&path)
+    }
+
+    for path in paths::ancestors(manifest_path, None).skip(2) {
+        if path.ends_with("target/package") {
+            break;
+        }
+
+        let ances_manifest_path = path.join("Cargo.toml");
+        if ances_manifest_path.exists() {
+            let man = parse_manifest(&ances_manifest_path, config)?;
+            match man.workspace_config() {
+                WorkspaceConfig::Root(ref ances_root_config) => {
+                    debug!("find_root - found a root checking exclusion");
+                    if !ances_root_config.is_excluded(manifest_path) {
+                        let mut inheritable_map = config.inheritable.borrow_mut();
+                        match inheritable_map.get(&ances_manifest_path) {
+                            Some(_) => (),
+                            None => {
```

Comment:
Is there a reason this only inserts into the cache and doesn't read from it?

(feel free to ignore if we end up removing the caching from this PR)